renku/core/incubation/database.py

```diff
@@ -0,0 +1,466 @@
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Custom database for store Persistent objects."""
+
+import datetime
+import hashlib
+import json
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple, Union
+from uuid import uuid4
+
+import BTrees.OOBTree
+from persistent import GHOST, UPTODATE, Persistent, PickleCache
+from persistent.interfaces import IPersistentDataManager
+from ZODB.interfaces import IConnection
+from ZODB.POSException import POSKeyError
+from ZODB.utils import z64
+from zope.interface import implementer
+
+MARKER = object()
+
+
+# NOTE These are used as _p_serial to mark if an object was read from storage or is new
+NEW = z64  # NOTE: Do not change this value since this is the default when a Persistent object is created
+PERSISTED = b"1" * 8
+
+MIN_COMPRESSED_FILENAME_LENGTH = 64
+
+
+@implementer(IConnection, IPersistentDataManager)
+class Database:
+    """The Metadata Object Database."""
+
+    def __init__(self, path: Union[str, Path], root_type_names: Tuple[str] = ("Activity", "Entity", "Plan")):
+        self._storage: Storage = Storage(path)
+        self._root_type_names: Tuple[str] = tuple(root_type_names)
+        self._cache = PickleCache(self)
+        # The pre-cache is used by get to avoid infinite loops when objects load their state
+        self._pre_cache: Dict[bytes, Persistent] = {}
+        # Objects added explicitly by add() or when serializing other objects. After commit they are moved to _cache.
+        self._objects_to_commit: Dict[bytes, Persistent] = {}
+        self._reader = ObjectReader(database=self)
+        self._writer = ObjectWriter(database=self)
+
+    @property
+    def root(self):
+        """Return the database root object."""
+        try:
+            return self.get(b"root")
+        except POSKeyError:
+            root = BTrees.OOBTree.BTree()
+
+            for root_type in self._root_type_names:
+                self._create_root_type(root, root_type)
+
+            self.add(root, b"root")
+            root._p_changed = 1
+            return root
+
+    @staticmethod
+    def _create_root_type(root, root_type):
+        if root_type in root:
+            return
+
+        object = BTrees.OOBTree.BTree()
+        object._p_oid = root_type.encode("ascii")
+        root[root_type] = object
+
+    @staticmethod
+    def new_oid():
```

Comment:
This probably also doesn't need to be a static method

Reply:
For some reason I saw the other one first. Same as below.